<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="description" content="Among the many tools used for GWAS and Genomic Prediction, there seems to equally as many file formats to navigate. Running multiple different algorithms then requires switching between file...">
        <meta name="keywords" content="">
        <link rel="icon" href="./favicon.ico">

        <title>How-to: Edit GWAS and Genomic Prediction File Formats - Computational approaches to biological problems</title>

        <!-- Stylesheets -->
        <link href="./theme/css/bootstrap.min.css" rel="stylesheet">
        <link href="./theme/css/fonts.css" rel="stylesheet">
        <link href="./theme/css/nest.css" rel="stylesheet">
        <link href="./theme/css/pygment.css" rel="stylesheet">
        <!-- /Stylesheets -->

        <!-- RSS Feeds -->
        <!-- /RSS Feeds -->

        <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
          <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
          <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
        <![endif]-->



    </head>

    <body>

        <!-- Header -->
    <div class="header-container gradient">

            <!-- Static navbar -->
            <div class="container">
                <div class="header-nav">
                    <div class="header-logo">
                        <a class="pull-left" href="./"><img class="mr20" src="./images/boat-logo_small.jpeg" alt="logo">Computational approaches to biological problems</a>
                    </div>
                    <div class="nav pull-right">
                                <a href="./">Home</a>
                                <a href="./about-me.html">About</a>
                                <a href="https://github.com/jlboat">GitHub</a>
                                <a href="https://scholar.google.com/citations?hl=en&user=q5G2bJEAAAAJ&view_op=list_works&gmla=AJsN-F4gfKvkjBIT6Nv8Xy97IIUHuhFi9jF8x5ntZveyPGo_xO2ws-_5NjCw-_9z9CD2mWJnlbL1DeMd-0l-IUNYPsQ1NR4wu8s_l6YjZYlsb9mTQ1d8TX8">Pubs</a>
                    </div>
                </div>
            </div>
            <!-- /Static navbar -->

            <!-- Header -->
    <!-- Header -->
    <div class="container header-wrapper">
        <div class="row">
              <div class="col-lg-12">
                  <div class="header-content">
                      <h1 class="header-title">How-to: Edit GWAS and Genomic Prediction File Formats</h1>
                      <p class="header-date">By <a href="./author/j-lucas-boatwright.html">J. Lucas Boatwright</a>, Thu 10 October 2019, in category <a href="./category/misc.html">Misc</a></p>
                      <div class="header-underline"></div>
                      <div class="clearfix"></div>
                      <p class="pull-right header-tags">
                          <span class="glyphicon glyphicon-tags mr5" aria-hidden="true"></span>
                      </p>
                  </div>
              </div>
        </div>
    </div>
    <!-- /Header -->
            <!-- /Header -->

        </div>
        <!-- /Header -->


        <!-- Content -->
    <div class="container content">
        <p>Among the many tools used for GWAS and Genomic Prediction, there seems to equally as many file formats to navigate. Running multiple different algorithms then requires switching between file formats, which can be a tedious if not time-absorbing task. For both my future sanity as well as yours, I've provided a brief overview of some common formats and recommended routes for file conversion.</p>
<p>To start, I highly recommend that all genotype files be originally generated in VCF format. If you can do so following the GATK best practices for variant calling, all the better. The reason I recommend this is because VCF files are generally the most descriptive of all the genotype formats and conversion from VCF format to others results in a loss of informaton.</p>
<p>Once you have a VCF file, genotype file conversion is relatively simple. You can use Tassel which provides options to convert among several genotype file formats:</p>
<p>VCF | HapMap | HapMap Diploid | Plink | HDF5 | Phylip (interleaved or sequential) | Table</p>
<p>This may be done using the Tassel GUI (graphical user interface) or on command line.</p>
<p>Alternatively, you can use vcftools to convert a VCF or BCF (binary VCF) to the following formats:</p>
<p>BEAGLE | IMPUTE | LDhat | Plink | Plink TPED (transposed) | Numeric Matrix</p>
<p>Conversion among genotype files is <em>typically</em> relatively simple but can be more complicated when converting between phased or unphased genotypes or when files with multi-allelic sites need to be converted to contain only bi-allelic sites.</p>
<p>As for phenotype formats, the main issue is typically determining what format you need for your program of choice and then figuring out how best to convert it.</p>
<p>Plink format for phenotypes has the following tab-delimited structure:</p>
<ul>
<li>FID = Family ID</li>
<li>IID = Individual ID</li>
</ul>
<p>In this example, I don't have FIDs so the IID is just repeated for both.</p>
<table>
<thead>
<tr>
<th>FID</th>
<th>IID</th>
<th>Phenotype1</th>
<th>...</th>
<th>PhenotypeN</th>
</tr>
</thead>
<tbody>
<tr>
<td>Indiv1</td>
<td>Indiv1</td>
<td>0</td>
<td>...</td>
<td>0</td>
</tr>
<tr>
<td>Indiv2</td>
<td>Indiv2</td>
<td>3</td>
<td>...</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>This is slightly different from the Tassel tab-delimited format for phenotypes:</p>
<table>
<thead>
<tr>
<th>\<Trait></th>
<th>Phenotype1</th>
<th>...</th>
<th>PhenotypeN</th>
</tr>
</thead>
<tbody>
<tr>
<td>Indiv1</td>
<td>0</td>
<td>...</td>
<td>0</td>
</tr>
<tr>
<td>Indiv2</td>
<td>3</td>
<td>...</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>Tassel will allow you to save a phenotype file into either Tassel or Plink format.</p>
<p>If you're running GAPIT, then the phenotype file is read into an R dataframe. That means that the delimiter used is less important since files can be read with "read.csv" or "read.table". However, you will generally want the file to looks like this:</p>
<table>
<thead>
<tr>
<th>Taxa</th>
<th>Phenotype1</th>
<th>...</th>
<th>PhenotypeN</th>
</tr>
</thead>
<tbody>
<tr>
<td>Indiv1</td>
<td>0</td>
<td>...</td>
<td>0</td>
</tr>
<tr>
<td>Indiv2</td>
<td>3</td>
<td>...</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>And, when you read the file, you want to designate that there are headers for the phenotype file ("head=TRUE"). Don't do this for the genotype file, though (in HapMap format). Even though there <em>is</em> a header, it should be read as "head=FALSE" so that the header occurs on the first row.</p>
<p>It's also worth noting that while the first 11 HapMap columns are required for GAPIT, only three of them are used ("rs" a.k.a. SNP name,  "chrom" and "pos"). So, the other eight columns may be filled with "NA".</p>
<p>Now, let's assume instead that you want to run Plink or GEMMA. GEMMA can use Plink file formats, so let's use that common format. Conversion from VCF to Plink files is easily acheived using Tassel or vcftools. </p>
<p>To begin, Plink requires both a PED (pedigree) and MAP (genetic map) file. Plink PED file format requires all markers be biallelic and the file look like so (header included here for clarity -- not in actual PED file):</p>
<table>
<thead>
<tr>
<th>FamilyID</th>
<th>IndividualID</th>
<th>PaternalID</th>
<th>MaternalID</th>
<th>Sex</th>
<th>Phenotype</th>
<th>Snp1</th>
<th>...</th>
<th>SnpN</th>
</tr>
</thead>
<tbody>
<tr>
<td>Indiv1</td>
<td>Indiv1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>A</td>
<td>...</td>
<td>G</td>
</tr>
<tr>
<td>Indiv2</td>
<td>Indiv2</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>A</td>
<td>...</td>
<td>G</td>
</tr>
</tbody>
</table>
<p>And the Plink MAP file looks like (again, header included for clarity only):</p>
<table>
<thead>
<tr>
<th>chromosome</th>
<th>SnpID</th>
<th>GeneticDistance</th>
<th>BasePairPosition</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>S01_3918</td>
<td>0</td>
<td>3918</td>
</tr>
<tr>
<td>1</td>
<td>S01_12215</td>
<td>0</td>
<td>12215</td>
</tr>
<tr>
<td>1</td>
<td>S01_23664</td>
<td>0</td>
<td>23664</td>
</tr>
</tbody>
</table>
<div class="highlight"><pre><span></span><span class="c1"># This will generate the PED and MAP files -- note: Only biallelic loci will be output</span>
<span class="c1"># Tassel can also generate the PED and MAP files, but Tassel retains multiallelic loci</span>
<span class="c1">#     which is a problem for GEMMA</span>
vcftools --vcf genotypes.vcf --plink --out genotypes 

<span class="c1"># This will generate BED (binary PED), NoSex (if --allow-no-sex is included), </span>
<span class="c1"># FAM (PLINK sample information file), BIM (PLINK extended MAP file)  and log files</span>
plink --file genotypes --pheno phenotypes.txt --allow-no-sex --make-bed --out genotypes
</pre></div>


<p>The BED, BIM and FAM files are referred to as the Plink binary fileset. These are what you want to run GEMMA and Plink.</p>
<p>The last style of genotype matrix typically seen is the numeric genotype matrix. Depending on the software, you may want this matrix in 0, 1, 2 format or in -1, 0, 1. This matrix format is useful in genomic prediction as a design matrix or during the calculation of a genomic relationship matrix.</p>
<p>Conversion to 012 is easily accomplished using vcftools.</p>
<div class="highlight"><pre><span></span>vcftools --vcf my.vcf --012 --out geno_matrix
</pre></div>


<p>Then, conversion to a -101 format simply requires that the matrix be read into either R or python and one be subtracted from the matrix. For example:</p>
<div class="highlight"><pre><span></span><span class="c1"># Commented code for a small-scale example</span>
<span class="c1"># myMatrix &lt;- matrix(c(c(0,1,2),c(2,1,0),c(1,2,0)), nrow=3)</span>

<span class="n">myMatrix</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;geno_matrix.012&quot;</span><span class="p">,</span> <span class="n">head</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="n">myMatrix</span> <span class="o">&lt;-</span> <span class="n">myMatrix</span> <span class="o">-</span> <span class="m">1</span>
</pre></div>


<p>Do note, however, that different programs require missing data in particular formats. Here, the 012 matrix has missing as -1 and the -101 has -2. So, change for NAs as necessary.</p>
<p>That should cover the majority of file conversions needed for performing GWAS or Genomic Prediction. Good luck!</p>


        
    </div>
        <!-- /Content --> 

        <!-- Footer -->
        <div class="footer gradient-2">
            <div class="container footer-container ">
                <div class="row">
                    <div class="col-xs-4 col-sm-3 col-md-3 col-lg-3">
                        <div class="footer-title"></div>
                        <ul class="list-unstyled">
                        </ul>
                    </div>
                    <div class="col-xs-4 col-sm-3 col-md-3 col-lg-3">
                    </div>
                    <div class="col-xs-4 col-sm-3 col-md-3 col-lg-3">
                        <div class="footer-title"></div>
                        <ul class="list-unstyled">
                            <li><a href="about-me.html" target="_blank">about-me</a></li>
                            <li><a href="https://github.com/jlboat" target="_blank">gitHub</a></li>
                        </ul>
                    </div> 
                    <div class="col-xs-12 col-sm-3 col-md-3 col-lg-3">
                        <p class="pull-right text-right">
                            <small><em>Proudly powered by <a href="http://docs.getpelican.com/" target="_blank">pelican</a></em></small><br/>
                            <small><em>Theme and code by <a href="https://github.com/molivier" target="_blank">molivier</a></em></small><br/>
                            <small></small>
                        </p>
                    </div>
                </div>
            </div>
        </div>
        <!-- /Footer -->
    </body>
</html>